piptools/build.py

```diff
@@ -0,0 +1,152 @@
+                     environment or to create an isolated one and invoke it
+                     there.
+    """
+
+    src_dir = os.path.dirname(os.path.abspath(src_file))
+    with _create_project_builder(src_dir, isolated) as builder:
+        metadata = _build_project_wheel_metadata(builder)
+        extras = tuple(metadata.get_all("Provides-Extra") or ())
+        requirements = tuple(
+            _prepare_requirements(metadata=metadata, src_file=src_file)
+        )
+        build_requirements = tuple(
+            _prepare_build_requirements(
+                builder=builder,
+                src_file=src_file,
+                distributions=build_distributions,
+                package_name=metadata.get_all("Name")[0],
+            )
+        )
+        return ProjectMetadata(
+            extras=extras,
+            requirements=requirements,
+            build_requirements=build_requirements,
+        )
+
+
+@contextlib.contextmanager
+def _create_project_builder(
+    src_dir: str, isolated: bool = False
+) -> Iterator[build.ProjectBuilder]:
+    builder = build.ProjectBuilder(
+        os.fspath(src_dir),
+        runner=pyproject_hooks.default_subprocess_runner,
+    )
+
+    if not isolated:
+        yield builder
+        return
+
+    with build.env.IsolatedEnvBuilder() as env:
+        builder.python_executable = env.executable
+        builder.scripts_dir = env.scripts_dir
+        env.install(builder.build_system_requires)
+        env.install(builder.get_requires_for_build("wheel"))
+        yield builder
+
+
+def _build_project_wheel_metadata(
+    builder: build.ProjectBuilder,
+) -> importlib_metadata.PackageMetadata:
+    with tempfile.TemporaryDirectory() as tmpdir:
+        path = pathlib.Path(builder.metadata_path(tmpdir))
+        return importlib_metadata.PathDistribution(path).metadata
+
+
+def _prepare_requirements(
+    metadata: importlib_metadata.PackageMetadata, src_file: str
+) -> Iterator[InstallRequirement]:
+    package_name = metadata.get_all("Name")[0]
+    comes_from = f"{package_name} ({src_file})"
+
+    for req in metadata.get_all("Requires-Dist") or []:
+        parts = parse_req_from_line(req, comes_from)
+        if parts.requirement.name == package_name:
+            package_dir = os.path.dirname(os.path.abspath(src_file))
+            # Replace package name with package directory in the requirement
+            # string so that pip can find the package as self-referential.
+            # Note the string can contain extras, so we need to replace only
+            # the package name, not the whole string.
+            replaced_package_name = req.replace(package_name, package_dir, 1)
+            parts = parse_req_from_line(replaced_package_name, comes_from)
+
+        yield InstallRequirement(
+            parts.requirement,
+            comes_from,
+            link=parts.link,
+            markers=parts.markers,
+            extras=parts.extras,
+        )
+
+
+def _prepare_build_requirements(
```

Comment:
What does prepare mean in this context?
Does the function have any side effects that I am missing?

Reply:
No side effects, it constricts an iterator of `InstallRequirement` instances (build requirements) based on a given build metadata.